src/client/iframe.ts

```diff
@@ -0,0 +1,101 @@
+import { PluginDevMode, PluginClient, PluginOptions } from './client'
+import { Message, ModuleProfile } from '../types'
+import { ApiMap, getApiMap, listenOnThemeChanged } from './api'
+
+/**
+ * Check if the sender has the right origin
+ * @param msgOrigin The origin of the incoming message
+ * @param devMode Devmode options
+ */
+function checkOrigin(msgOrigin: string, devMode?: PluginDevMode) {
+  const localhost = devMode ? [
+    `http://127.0.0.1:${devMode.port}`,
+    `http://localhost:${devMode.port}`,
+    `https://127.0.0.1:${devMode.port}`,
+    `https://localhost:${devMode.port}`,
+  ] : []
+  return [
+    "http://remix-alpha.ethereum.org",
+    "http://remix.ethereum.org",
+    "https://remix-alpha.ethereum.org",
+    "https://remix.ethereum.org",
+    ...localhost
+  ].includes(msgOrigin)
+}
+
+/**
+ * Start listening on the IDE though PostMessage
+ * @param store A store to put the messages into
+ */
+function connectIframe(client: PluginClient) {
+  let loaded = false
+
+  async function getMessage(event: MessageEvent) {
+    if (!(event.source)) throw new Error('No source')
+
+    // Check that the origin is the right one
+    const devMode = client.devMode
+    if (!checkOrigin(event.origin, devMode)) return
+
+    // Get the data
+    if (!event.data) throw new Error('No data')
+    const { action, key, name, payload, id, requestInfo, error } = event.data as Message
+
+    try {
+      // If handshake set loaded
+      if (action === 'request' && key === 'handshake') {
+        loaded = true
+        client.events.emit('loaded')
+        client.events.on('send', (message: Message) => {
+          event.source.postMessage(message, event.origin as any)
+        })
+      }
+
+      // Check if is loaded
+      if (!loaded) throw new Error('Handshake before communicating')
+
+      switch (action) {
+        case 'notification': {
+          client.events.emit(`[${name}] ${key}`, ...payload)
```

Comment:
interesting way of providing separation

Reply:
indeed

Reply:
should that one use https://github.com/ethereum/remix-plugin/pull/68/files#diff-2609dc589ff34d954b51407ff2140828R32 ?